### PR TITLE
KAFKA-15399: Enable OffloadAndConsumeFromLeader test

### PR DESCRIPTION
**Pull request by kamalcph**
Co-authored-by: Alexandre Dupriez [alexandre.dupriez@gmail.com](mailto:alexandre.dupriez@gmail.com)
Co-authored-by: Kamal Chandraprakash [kamal.chandraprakash@gmail.com](mailto:kamal.chandraprakash@gmail.com)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by kamalcph** [comment]
### Test Report

```scala

[SUCCESS] (1) create topic: Topic[name=topicA partition-count=1 replication-factor=1 segment-size=1 assignment=null properties={remote.storage.enable=true, local.retention.bytes=1, index.interval.bytes=1, segment.index.bytes=12}]

[SUCCESS] (2) produce-records: topicA-0
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k3, value=v3, timestamp=null)
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=0 record-count=1]
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=1 record-count=1]

[SUCCESS] (3) create topic: Topic[name=topicB partition-count=1 replication-factor=1 segment-size=2 assignment=null properties={remote.storage.enable=true, local.retention.bytes=1, index.interval.bytes=1, segment.index.bytes=24}]

[SUCCESS] (4) produce-records: topicB-0
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k3, value=v3, timestamp=null)
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k4, value=v4, timestamp=null)
    ProducerRecord(topic=topicB, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k5, value=v5, timestamp=null)
    Segment[partition=topicB-0 offloaded-by-broker-id=0 base-offset=0 record-count=2]
    Segment[partition=topicB-0 offloaded-by-broker-id=0 base-offset=2 record-count=2]

[SUCCESS] (5) consume-action:
  topic-partition = topicA-0
  fetch-offset = 1
  expected-record-count = 2
  expected-record-from-tiered-storage = 1

[SUCCESS] (6) consume-action:
  topic-partition = topicB-0
  fetch-offset = 1
  expected-record-count = 4
  expected-record-from-tiered-storage = 3

Content of local tiered storage:

Broker IDFile                                       | Offsets |      Records 
----------------------------------------------------------------------------------------
topicA-0                                            |         |              
    00000000000000000000-z5vzX2egQSWFoceLXBFIpw.log |       0 |     (k1, v1) 
                                                    |         |              
    00000000000000000001-f33sDWOvTCWlMkiwxlbo4Q.log |       1 |     (k2, v2) 
                                                    |         |              
topicB-0                                            |         |              
    00000000000000000000-iCzBGsRPRuuGfYY3XqCNgg.log |       0 |     (k1, v1) 
                                                    |       1 |     (k2, v2) 
                                                    |         |              
    00000000000000000002-MQelU_s8SvO2mC_BcFqDBg.log |       2 |     (k3, v3) 
                                                    |       3 |     (k4, v4) 
                                                    |         |              
```

File: core/src/test/scala/integration/kafka/admin/RemoteTopicCrudTest.scala [hunk 1]
```diff
@@ -365,12 +363,11 @@ object MyRemoteStorageManager {
   val deleteSegmentEventCounter = new AtomicInteger(0)
 }
 
-class MyRemoteStorageManager extends NoOpRemoteStorageManager with Logging {
+class MyRemoteStorageManager extends NoOpRemoteStorageManager {
   import MyRemoteStorageManager._
 
   override def deleteLogSegmentData(remoteLogSegmentMetadata: RemoteLogSegmentMetadata): Unit = {
     deleteSegmentEventCounter.incrementAndGet()
-    info(s"Deleted the remote log segment: $remoteLogSegmentMetadata, counter: ${deleteSegmentEventCounter.get()}")
```
Comment (satishd) [inline core/src/test/scala/integration/kafka/admin/RemoteTopicCrudTest.scala:373]: Why is this logging removed?

**Comment by kamalcph** [comment]
@satishd @divijvaidya @showuon @clolov 

Call for review. PTAL.

File: storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java [hunk 2]
```diff
@@ -138,18 +140,21 @@ private boolean compare(ByteBuffer lhs,
     private SimpleRecord convert(Object recordCandidate) {
         if (recordCandidate instanceof ProducerRecord) {
             ProducerRecord<?, ?> record = (ProducerRecord<?, ?>) recordCandidate;
+            long timestamp = record.timestamp() == null ? RecordBatch.NO_TIMESTAMP : record.timestamp();
             ByteBuffer keyBytes =
                     Utils.wrapNullable(keySerde.serializer().serialize(topicPartition.topic(), (K) record.key()));
             ByteBuffer valueBytes =
                     Utils.wrapNullable(valueSerde.serializer().serialize(topicPartition.topic(), (V) record.value()));
-            return new SimpleRecord(record.timestamp(), keyBytes, valueBytes, record.headers().toArray());
+            Header[] headers = record.headers() != null ? record.headers().toArray() : Record.EMPTY_HEADERS;
```
Comment (satishd) [inline storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java:148]: nit: may introduce `getHeaders(record)` to return headers as it is used at other places. 
```
Header[] getHeaders(ProducerRecord record) {
    return record.headers() != null ? record.headers().toArray() : Record.EMPTY_HEADERS;
}
```

**Review by satishd — COMMENTED** [review]
Thanks @kamalcph for the PR, left a couple of minor comments.

File: core/src/test/scala/integration/kafka/admin/RemoteTopicCrudTest.scala (hunk 1 above)
Comment (kamalcph) [reply core/src/test/scala/integration/kafka/admin/RemoteTopicCrudTest.scala:373]: Previously, we were checking it manually by counting the number of log statements. Fixed the test by adding an assertion which validates the number of deleted segments.

File: storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java (hunk 2 above)
Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java:148]: We have three different records: ProducerRecord, ConsumerRecord and SimpleRecord. They don't extend a common interface.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java [hunk 3]
```diff
@@ -127,10 +127,10 @@ protected void writeTestSpecifications(TieredStorageTestBuilder builder) {
                  *       - For topic B, only one segment is present in the tiered storage, as asserted by the
                  *         previous sub-test-case.
                  */
-                .bounce(broker)
+                // .bounce(broker)
```
Comment (kamalcph) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/OffloadAndConsumeFromLeaderTest.java:130]: Bouncing the broker updates the log-start-offset to first-local-log-segment-base-offset in [LogLoader](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/LogLoader.scala#L183), so the remote log segments are removed due to breach by log-start-offset. Will fix this issue as part of KAFKA-15351.

File: storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java (hunk 2 above)
Comment (satishd) [reply storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java:148]: okay, makes sense.

File: storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java [hunk 4]
```diff
@@ -138,18 +140,21 @@ private boolean compare(ByteBuffer lhs,
     private SimpleRecord convert(Object recordCandidate) {
         if (recordCandidate instanceof ProducerRecord) {
             ProducerRecord<?, ?> record = (ProducerRecord<?, ?>) recordCandidate;
+            long timestamp = record.timestamp() == null ? RecordBatch.NO_TIMESTAMP : record.timestamp();
```
Comment (clolov) [inline storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java]: Nit: Can either all conditions be == null or != null, but not mixed?

File: storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java [hunk 5]
```diff
@@ -138,18 +140,21 @@ private boolean compare(ByteBuffer lhs,
...
+            return new SimpleRecord(timestamp, keyBytes, valueBytes, headers);
         } else if (recordCandidate instanceof ConsumerRecord) {
             ConsumerRecord<?, ?> record = (ConsumerRecord<?, ?>) recordCandidate;
             ByteBuffer keyBytes =
                     Utils.wrapNullable(keySerde.serializer().serialize(topicPartition.topic(), (K) record.key()));
             ByteBuffer valueBytes =
                     Utils.wrapNullable(valueSerde.serializer().serialize(topicPartition.topic(), (V) record.value()));
-            return new SimpleRecord(record.timestamp(), keyBytes, valueBytes, record.headers().toArray());
+            Header[] headers = record.headers() != null ? record.headers().toArray() : Record.EMPTY_HEADERS;
+            return new SimpleRecord(record.timestamp(), keyBytes, valueBytes, headers);
```
Comment (clolov) [inline storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java:157]: For my understanding, why do you not need to carry out the null check for the timestamp of the record here, but you had to carry it above?

File: core/src/test/scala/integration/kafka/admin/RemoteTopicCrudTest.scala [hunk 6]
```diff
@@ -293,12 +294,9 @@ class RemoteTopicCrudTest extends IntegrationTestHarness {
...
     assertThrowsException(classOf[UnknownTopicOrPartitionException],
       () => TestUtils.describeTopic(createAdminClient(), testTopicName), "Topic should be deleted")
-
-    // FIXME: It seems the storage manager is being instantiated in different class loader so couldn't verify the value
-    //  but ensured it by adding a log statement in the storage manager (manually).
-    //    assertEquals(numPartitions * MyRemoteLogMetadataManager.segmentCount,
-    //      MyRemoteStorageManager.deleteSegmentEventCounter.get(),
-    //      "Remote log segments should be deleted only once by the leader")
+    TestUtils.waitUntilTrue(() =>
+      numPartitions * MyRemoteLogMetadataManager.segmentCount == MyRemoteStorageManager.deleteSegmentEventCounter.get(),
```
Comment (clolov) [inline core/src/test/scala/integration/kafka/admin/RemoteTopicCrudTest.scala]: Can you change the `segmentCount` property of MyRemoteLogMetadataManager to `segmentCountPerPartition`?

File: storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java [hunk 7]
```diff
@@ -307,7 +307,7 @@ public Optional<CustomMetadata> copyLogSegmentData(final RemoteLogSegmentMetadat
             throws RemoteStorageException {
         Callable<Optional<CustomMetadata>> callable = () -> {
             final RemoteLogSegmentId id = metadata.remoteLogSegmentId();
```
Comment (clolov) [inline storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java]: On line 316 you can change `id.topicIdPartition()` to `metadata.topicIdPartition()` and remove this line

File: storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java [hunk 8]
```diff
@@ -309,7 +310,5 @@ public void printReport(PrintStream output) {
 
     @Override
     public void close() throws IOException {
-        Utils.closeAll(producer, consumer);
```
Comment (clolov) [inline storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:312]: For my understanding, why are these safe to remove?

**Review by clolov — COMMENTED** [review]
Thanks for this! Carried out a first round of review

File: storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java (hunk 8 above)
Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:312]: clients are created using harness (harness.createConsumer), so they are auto-closed on tear-down. This is one of the follow-up [comment](https://github.com/apache/kafka/pull/14116#discussion_r1294668715) from PR #14116.

File: storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java (hunk 5 above)
Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/utils/RecordsKeyValueMatcher.java:157]: ProducerRecord.timestamp() returns object Long whereas the other records return primitive long.

File: storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java (hunk 8 above)
Comment (clolov) [reply storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:312]: Okay, do you need the overriden method if it has no implementation?

Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestContext.java:312]: Left it to close any other items in future if required.

**Review by divijvaidya — APPROVED** [review]
Looks good. I like the additional refactoring/ cleanup performed as part of this PR. Thank you for that.

Since this test is not able to run on CI due to thread leaks by other tests, I will be willing to merge it after ensuring that the test is successful locally. The thread leaks are associated with some other tests.

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for addressing the comments. LGTM. 

This test is failed in the jenkins jobs because of the expected threads did not close by other tests. I was able to run successfully multiple times in my local env. So, I am fine with these changes.

```
➜  kafka git:(KAFKA-15399) ./gradlew storage:test --tests org.apache.kafka.tiered.storage.integration.OffloadAndConsumeFromLeaderTest

> Configure project :
Starting build with version 3.7.0-SNAPSHOT (commit id 60404bb8) using Gradle 8.2.1, Java 11 and Scala 2.13.11
Build properties: maxParallelForks=12, maxScalacThreads=8, maxTestRetries=0

> Task :storage:test

Gradle Test Run :storage:test > Gradle Test Executor 29 > OffloadAndConsumeFromLeaderTest > executeTieredStorageTest() PASSED

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 34s
```

**Comment by divijvaidya** [comment]
@clolov has a national holiday today. Merging this in since we have 2 approvals already and merging this test is time sensitive due to code freeze. We can address his comments once he is back in a separate PR.

**Comment by divijvaidya** [comment]
Merging into 3.6

**Pull request merged**